src/base/bittorrent/private/filterparserthread.cpp

```diff
@@ -123,34 +129,49 @@ int FilterParserThread::parseDATFilterFile()
     }
 
     static const int bufferSize = 2 * 1024 * 1024; // 2 MiB
-    char *buffer = new char[bufferSize];
+    std::vector<char> buffer(bufferSize, 0); // seems a bit faster than QVector
     qint64 bytesRead = 0;
     int offset = 0;
     int start = 0;
     int endOfLine = -1;
     int nbLine = 0;
 
     while (true) {
-        memset(buffer + offset, 0, bufferSize - offset);
-        bytesRead = file.read(buffer + offset, bufferSize - offset);
-        if (bytesRead < 1)
+        bytesRead = file.read(buffer.data() + offset, bufferSize - offset);
+        if (bytesRead < 0)
             break;
         int dataSize = bytesRead + offset;
+        if (bytesRead == 0 && dataSize == 0)
+            break;
 
         for (start = 0; start < dataSize; ++start) {
             endOfLine = -1;
-            for (int i = start; i < dataSize; ++i) {
-                if (buffer[i] == '\n') {
-                    endOfLine = i;
-                    break;
+            // The file might have ended without the last line having a newline
+            if (!(bytesRead == 0 && dataSize > 0)) {
+                for (int i = start; i < dataSize; ++i) {
+                    if (buffer[i] == '\n') {
+                        endOfLine = i;
+                        // We need to NULL the newline in case the line has only an IP range.
+                        // In that case the parser won't work for the end IP, because it ends
+                        // with the newline and not with a number.
+                        buffer[i] = '\0';
+                        break;
+                    }
                 }
             }
+            else {
+                if (dataSize < bufferSize)
+                    buffer[dataSize] = '\0';
+                else
+                    buffer.push_back('\0');
```

Comment:
This is a worst case scenario and will trigger only one realloctions during each parsing of the "faulty" file.
I don't consider this scenario to be prevalent.

Do you have any other suggestion?